File: snippets/rust/src/list_payments.rs
use std::sync::Arc;

use anyhow::Result;
use breez_sdk_liquid::prelude::*;

async fn get_payment(sdk: Arc<LiquidSdk>) -> Result<Option<Payment>> {
    // ANCHOR: get-payment
    let payment_hash = "<payment hash>".to_string();
    let payment = sdk.get_payment(&GetPaymentRequest::Lightning {
        payment_hash
    }).await?;
    // ANCHOR_END: get-payment

    Ok(payment)
}

async fn list_payments(sdk: Arc<LiquidSdk>) -> Result<Vec<Payment>> {
    // ANCHOR: list-payments
    let payments = sdk.list_payments(&ListPaymentsRequest::default()).await?;
    // ANCHOR_END: list-payments

    Ok(payments)
}

async fn list_payments_filtered(sdk: Arc<LiquidSdk>) -> Result<Vec<Payment>> {
    // ANCHOR: list-payments-filtered
    let payments = sdk
        .list_payments(&ListPaymentsRequest {
            filters: Some(vec![PaymentType::Send]),
            from_timestamp: Some(1696880000),
            to_timestamp: Some(1696959200),
            offset: Some(0),
            limit: Some(50),
            details: None,
        })
        .await?;
    // ANCHOR_END: list-payments-filtered

    Ok(payments)
}

async fn list_payments_details_address(sdk: Arc<LiquidSdk>) -> Result<Vec<Payment>> {
    // ANCHOR: list-payments-details-address
    let address = "<Bitcoin address>".to_string();
    let payments = sdk
        .list_payments(&ListPaymentsRequest {
            filters: None,
            from_timestamp: None,
            to_timestamp: None,
            offset: None,
            limit: None,
            details: Some(ListPaymentDetails::Bitcoin {
                address,
            }),            
        })
        .await?;
    // ANCHOR_END: list-payments-details-address

    Ok(payments)
}

async fn list_payments_details_destination(sdk: Arc<LiquidSdk>) -> Result<Vec<Payment>> {
    // ANCHOR: list-payments-details-destination
    let destination = "<Liquid BIP21 or address>".to_string();
    let payments = sdk
        .list_payments(&ListPaymentsRequest {
            filters: None,
            from_timestamp: None,
            to_timestamp: None,
            offset: None,
            limit: None,
            details: Some(ListPaymentDetails::Liquid {
                destination,
            }),            
        })
        .await?;
    // ANCHOR_END: list-payments-details-addrdestinationess

    Ok(payments)
}

File: snippets/go/list_payments.go
package example

import (
	"log"

	"github.com/breez/breez-sdk-liquid-go/breez_sdk_liquid"
)

func GetPayment(sdk *breez_sdk_liquid.BindingLiquidSdk) {
	// ANCHOR: get-payment
	paymentHash := "<payment hash>"
	req := breez_sdk_liquid.GetPaymentRequestLightning{
		PaymentHash: paymentHash,
	}
	if payment, err := sdk.GetPayment(req); err == nil {
		log.Printf("%#v", payment)
	}
	// ANCHOR_END: get-payment
}

func ListPayments(sdk *breez_sdk_liquid.BindingLiquidSdk) {
	// ANCHOR: list-payments
	if payments, err := sdk.ListPayments(breez_sdk_liquid.ListPaymentsRequest{}); err == nil {
		log.Printf("%#v", payments)
	}
	// ANCHOR_END: list-payments
}

func ListPaymentsFiltered(sdk *breez_sdk_liquid.BindingLiquidSdk) {
	// ANCHOR: list-payments-filtered
	filters := []breez_sdk_liquid.PaymentType{breez_sdk_liquid.PaymentTypeSend}
	fromTimestamp := int64(1696880000)
	toTimestamp := int64(1696959200)
	limit := uint32(50)
	offset := uint32(0)
	listPaymentsRequest := breez_sdk_liquid.ListPaymentsRequest{
		Filters:       &filters,
		FromTimestamp: &fromTimestamp,
		ToTimestamp:   &toTimestamp,
		Offset:        &offset,
		Limit:         &limit,
	}
	if payments, err := sdk.ListPayments(listPaymentsRequest); err == nil {
		log.Printf("%#v", payments)
	}
	// ANCHOR_END: list-payments-filtered
}

func ListPaymentsDetailsAddress(sdk *breez_sdk_liquid.BindingLiquidSdk) {
	// ANCHOR: list-payments-details-address
	address := "<Bitcoin address>"
	var details breez_sdk_liquid.ListPaymentDetails = breez_sdk_liquid.ListPaymentDetailsBitcoin{Address: address}
	listPaymentsRequest := breez_sdk_liquid.ListPaymentsRequest{
		Details: &details,
	}
	if payments, err := sdk.ListPayments(listPaymentsRequest); err == nil {
		log.Printf("%#v", payments)
	}
	// ANCHOR_END: list-payments-details-address
}

func ListPaymentsDetailsDestination(sdk *breez_sdk_liquid.BindingLiquidSdk) {
	// ANCHOR: list-payments-details-destination
	destination := "<Liquid BIP21 or address>"
	var details breez_sdk_liquid.ListPaymentDetails = breez_sdk_liquid.ListPaymentDetailsLiquid{Destination: destination}
	listPaymentsRequest := breez_sdk_liquid.ListPaymentsRequest{
		Details: &details,
	}
	if payments, err := sdk.ListPayments(listPaymentsRequest); err == nil {
		log.Printf("%#v", payments)
	}
	// ANCHOR_END: list-payments-details-destination
}
